newapi/ooniapi/auth.py

```diff
@@ -0,0 +1,213 @@
+"""
+Authentication API
+"""
+from datetime import datetime, timedelta
+from email.headerregistry import Address
+from email.message import EmailMessage
+from urllib.parse import urljoin
+import hashlib
+import re
+import smtplib
+
+from flask import Blueprint, current_app, request, make_response
+from flask.json import jsonify
+import jwt  # debdeps: python3-jwt
+
+from ooniapi.config import metrics
+from ooniapi.utils import cachedjson
+
+auth_blueprint = Blueprint("auth_api", "auth")
+
+"""
+Browser authentication - see probe_services.py for probe authentication
+Requirements:
+  - Never store users email address nor IP addresses nor passwords
+  - Verify email to limit spambots. Do not use capchas
+  - Support multiple sessions / devices, ability to register/login again
+
+Workflow:
+  Explorer:
+    - call register_user using an email and receive a temporary login link
+    - call login_user and receive a long-lived cookie
+    - call <TODO> using the previous email to get a new temp. login link
+    - call the citizenlab CRUD entry points using the cookie
+    - call bookmarked searches/urls/msmts entry points using the cookie
+"""
+
+# Courtesy of https://emailregex.com/
+EMAIL_RE = re.compile(r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)")
+
+
+def jerror(msg, code=400):
+    return make_response(jsonify(error=msg), code)
+
+
+def create_jwt(payload: dict) -> str:
+    key = current_app.config["JWT_ENCRYPTION_KEY"]
+    return jwt.encode(payload, key, algorithm="HS256").decode()
+
+
+def decode_jwt(token, **kw):
+    key = current_app.config["JWT_ENCRYPTION_KEY"]
+    return jwt.decode(token, key, algorithms=["HS256"], **kw)
+
+
+def hash_email_address(email_address: str) -> str:
+    key = current_app.config["JWT_ENCRYPTION_KEY"].encode()
```

Comment:
Are we sure we want to be using the same key for both the keyed hash function and the signing of the JWT tokens? I haven't looked at it carefully, but I wonder if by doing so we are in some way reducing the security of the key used.